archivist/runner.py

```diff
@@ -82,113 +109,101 @@ def run_steps(self, config: Dict):
         """Runs a step given parameters and the type of step.
 
         Args:
             step (dict): the steps map.
         """
 
         s = step["step"]
+        description = s.get("description")
+        if description is not None:
+            LOGGER.info(description)
+
         action_name = s.get("action")
         if action_name is None:
             raise ArchivistInvalidOperationError("Missing Action")
 
-        action = self._actions.get(action_name)
+        action, keywords = self._actions.get(action_name, (None, None))
         if action is None:
-            raise ArchivistInvalidOperationError(f"Illegal Action {action_name}")
-
-        description = s.get("description")
-        if description is not None:
-            LOGGER.info(description)
+            raise ArchivistInvalidOperationError(f"Illegal Action '{action_name}'")
 
-        # wait for a number of seconds
-        wait_time = s.get("wait_time", 0)
-        if wait_time > 0:
-            LOGGER.info("Waiting for %d seconds", wait_time)
-            time_sleep(wait_time)
-
-        # sort out argumenst to action method.
+        # sort out arguments to action method.
         # XXXX: add inspect.signature calls to validate ...
         # (at the moment we will get an upstream error if the step is incorrect)
         args = []
         asset_name = s.get("asset_name")
+        if asset_name is not None:
+            asset_id = self.asset_id(asset_name)
+            if asset_id is None:
+                raise ArchivistInvalidOperationError(f"Unknown Asset '{asset_name}'")
 
-        asset_id = self.asset_id(asset_name)
-        if asset_id is not None:
-            args.append(asset_id)
+            if "asset_id" not in keywords:
+                args.append(asset_id)
 
         kwargs = {}
-        data = {k: v for k, v in step.items() if k not in ("step", "confirm")}
+        keys = [
+            "step",
+        ]
+        if len(keywords) > 0:
+            keys.extend(keywords)
+            if "asset_id" in keywords:
+                kwargs["asset_id"] = asset_id
+
+            for k in keywords:
+                if k in step:
+                    kwargs[k] = step[k]
+
+        LOGGER.debug("keys %s", keys)
+        data = {k: v for k, v in step.items() if k not in keys}
         if data:
             args.append(data)
-        LOGGER.debug("args %s", args)
 
-        if "confirm" in step:
-            kwargs["confirm"] = step["confirm"]
+        # wait for a number of seconds
+        wait_time = s.get("wait_time", 0)
+        if wait_time > 0:
+            LOGGER.info("Waiting for %d seconds", wait_time)
+            time_sleep(wait_time)
+
+        LOGGER.debug("args %s", args)
+        if len(kwargs) > 0:
             LOGGER.debug("kwargs %s", kwargs)
             response = action(*args, **kwargs)
         else:
             response = action(*args)
 
+        # Some actions return a tuple
```

Comment:
Roll on dataclasses....

Once we have dataclasses the create() methods will either accept a dataclass or a dict. The additional kwargs will only represent args relating to PySDK itself (such as confirm=True) 

whilst this will not necessarily simplify this code it **will** make more easily explainable.

Reply:
OK I can see that helping. So I assume this will only last a short while after removing support for 3.6? If so then I'm not too worried and we can resolve this comment.

Reply:
this has been refactored into classes...